extensions/windowgram/init.lua

```diff
@@ -0,0 +1,210 @@
+
+local windowgram = {}
+
+-- Returns a table with the relative cols and rows.
+function parseWindowgram(wg)
+  local gridw = nil
+
+  local grid = {}
+
+  -- Split into lines
+  local lines = {}
+  for line in string.gmatch(wg, "[^\r\n]+") do
+    if #line > 0 then
+      table.insert(lines, line)
+    end
+  end
+  lines = cleanLines(lines)
+
+  for i, line in ipairs(lines) do
+    if gridw then
+      if gridw ~= line:len() then
+        error('inconsistent grid width in windowgram')
+      end
+    else
+      gridw=line:len()
+    end
+    for column = 1, #line do
+      local char = line:sub(column, column)
+      if not grid[char] then
+        -- new window, create it with size 1x1
+        grid[char] = { x = column , y = i}
+      else
+        -- expand it
+        grid[char].w = column
+        grid[char].h = i
+      end
+    end
+  end
+  return grid
+
+end
+
+function cleanLines(wg)
+  local target = {}
+  for i,l in ipairs(wg) do
+    -- Remove comments and whitespace
+    l = l:gsub('#.*','')
+    l = l:gsub("%s+","")
+    table.insert(target, l)
+  end
+  return target
+end
+
+--- hs.windowgram.getRegionRatios(regions)
+--- Function
+--- Gets a table with the regions as percentages of the screen resolution
+---
+--- Parameters:
+---  * regions - The regions table returned by windowgram.convertToRegions(wg)
+---
+--- Returns:
+---  * Table with ratios. ex; { x=0, y=0.5, w=0.5, h=1 }
+---
+--- Notes:
+--- * These are numbers between 0 and 1 and represent two points; the top left
+--- corner and the bottom right corner.
+function windowgram.getRegionRatios(regions)
+  local ratios = {}
+  local screenSize = hs.screen.mainScreen():frame()
+  for k,region in pairs(regions) do
+    ratios[k] = {}
+    ratios[k].x = region.x ~= 0 and region.x / screenSize.w or 0
+    ratios[k].y = region.y ~= 0 and region.y / screenSize.h or 0
+    ratios[k].w = region.w ~= 0 and region.w / screenSize.w or 0
+    ratios[k].h = region.h ~= 0 and region.h / screenSize.h or 0
+  end
+  return ratios
+end
+
+--- hs.windowgram.convertToRegions(wg)
+--- Function
+--- Gets a table with pixel coords describing a region, given a windowgram
+---
+--- Parameters:
+---  * wg - The windowgram: multiline string with the desired window layout
+---
+--- Returns:
+---  * Table with regions which contain an X, Y, W, and H values
+---
+--- Notes:
+---  * Regions are pixel based coordinates relative to the windowgram
+---
+--- ~~~lua
+--- local wg = [[
+---   AAAAAAAAAAAABBBBBBBBBBBB
+---   AAAAAAAAAAAABBBBBBBBBBBB
+---   AAAAAAAAAAAABBBBBBBBBBBB
+---   CCCCCCCCCCCCCCCCCCCCCCCC
+---   CCCCCCCCCCCCCCCCCCCCCCCC
+--- ]]
+--- local windows = windowgram.getWindows(wg)
+--- ~~~
+---
+--- On a 1920x1080 resolution, This will return the following table;
+--- ~~~lua
+--- print(wg["A"])
+--- -- prints { x=0, y=0, w=960, h=648 }
+--- print(wg["B"])
+--- -- prints { x=960, y=0, w=960, h=648 }
+--- print(wg["C"])
+--- -- prints { x=0, y=648, w=1920, h=432 }
+--- ~~~
```

Comment:
I kinda feel like this example, and ones for the other functions, should move up to the top level of the module, showing typical usage, broken down into its various stages.